### PR TITLE
Improve operator tracing and make it E2E work

### PR DESCRIPTION
**Pull request by xiaoxmeng**
Differential Revision: D64946367

This PR improves the operator trace framework and replay tool implementation:
(1) Let driver execution framework to handle the trace input collection and summary file generation.
The finish trace is either called when trace limit exceeds or on operator close instead of no
more input. So it can capture more information in summary like peak memory usage. By removing 
the trace input collection into the driver framework it eases the trace input collection for a spilling
operator. We add to capture the peak memory and input rows in summary so it helps identify the
hot operator or task on replay debugging.
(2) Change the trace storage layout to have root with query id for traces from different tasks
(3) Abstract the replay tool function into TraceReplayRunner class and derive in Meta internal code repo
to handle the Meta internal env setup and keep most common part in  TraceReplayRunner for OSS
(4) A couple of fixes in trace replay tool to make it E2E function in Meta for table writer use case
(5) A couple of file/class renaming to make the file/class name to be more specific as currently we only
support operator level trace collection and replay

### CODE REVIEW

**Review by duanmeng — APPROVED** [review]
@xiaoxmeng LGTM, thanks for your refactor.

File: velox/exec/Driver.cpp [hunk 1]
```diff
@@ -95,8 +95,8 @@ const core::QueryConfig& DriverCtx::queryConfig() const {
   return task->queryCtx()->queryConfig();
 }
 
-const std::optional<trace::QueryTraceConfig>& DriverCtx::traceConfig() const {
-  return task->queryTraceConfig();
+const std::optional<trace::TraceConfig>& DriverCtx::traceConfig() const {
+  return task->TraceConfig();
```
Comment (tanjialiang) [inline velox/exec/Driver.cpp:99]: lower case 't' please

File: velox/tool/trace/TraceReplayer.cpp [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/tool/trace/TraceReplayRunner.h"
+
+#include <gflags/gflags.h>
+
+int main(int argc, char** argv) {
```
Comment (tanjialiang) [inline velox/tool/trace/TraceReplayer.cpp:21]: Can we rename this file TraceReplayerMain?

File: velox/tool/trace/TraceReplayRunner.cpp [hunk 3]
```diff
@@ -155,76 +116,173 @@ std::unique_ptr<tool::trace::OperatorReplayerBase> createReplayer() {
...
 }
 } // namespace
 
-int main(int argc, char** argv) {
-  if (argc == 1) {
-    gflags::ShowUsageWithFlags(argv[0]);
-    return -1;
-  }
+TraceReplayRunner::TraceReplayRunner()
+    : ioExecutor_(std::make_unique<folly::IOThreadPoolExecutor>(
```
Comment (tanjialiang) [inline velox/tool/trace/TraceReplayRunner.cpp:219]: Can we add a name to this io pool? It will facilitate debugging

File: velox/exec/TraceUtil.h [hunk 4]
```diff
@@ -29,6 +30,52 @@ namespace facebook::velox::exec::trace {
 /// Creates a directory to store the query trace metdata and data.
 void createTraceDirectory(const std::string& traceDir);
 
+/// Returns the trace directory for a given query.
+std::string getQueryTraceDirectory(
```
Comment (tanjialiang) [inline velox/exec/TraceUtil.h:34]: s/get/compose/

File: velox/exec/TraceUtil.h [hunk 5]
```diff
@@ -29,6 +30,52 @@ namespace facebook::velox::exec::trace {
...
+    const std::string& traceDir,
+    const std::string& queryId);
+
+/// Returns the trace directory for a given query task.
+std::string getTaskTraceDirectory(
```
Comment (tanjialiang) [inline velox/exec/TraceUtil.h:39]: ditto, and below

File: velox/exec/TraceUtil.cpp [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+#include <folly/json.h>
+
+#include "velox/common/base/Exceptions.h"
+#include "velox/common/file/File.h"
+#include "velox/common/file/FileSystems.h"
+#include "velox/exec/Trace.h"
+
+namespace facebook::velox::exec::trace {
+namespace {
+std::string foundLastPathNode(const std::string& path) {
```
Comment (tanjialiang) [inline velox/exec/TraceUtil.cpp:28]: s/found/find/

File: velox/exec/PartitionedOutput.cpp [hunk 7]
```diff
@@ -233,8 +233,6 @@ void PartitionedOutput::estimateRowSizes() {
 }
 
 void PartitionedOutput::addInput(RowVectorPtr input) {
-  traceInput(input);
```
Comment (tanjialiang) [inline velox/exec/PartitionedOutput.cpp:236]: where do we trace it after removing this line?

Comment (tanjialiang) [reply velox/exec/PartitionedOutput.cpp:236]: nvm, saw it in Driver.cpp

**Review by tanjialiang — APPROVED** [review]
Thanks for the refactoring, LGTM

File: velox/exec/TraceUtil.h (hunk 4 above)
Comment (xiaoxmeng) [reply velox/exec/TraceUtil.h:34]: I will keep it as get

**Pull request closed without merging**